tests/fuzz/s2n_select_server_cert_test.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/x509v3.h>
+#include <openssl/x509.h>
+#include <openssl/pem.h>
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_random.h"
+#include "utils/s2n_safety.h"
+#include "s2n_test.h"
+
+#define MAX_TOKENS 256
+
+static void s2n_fuzz_atexit()
+{
+    s2n_cleanup();
+}
+
+int LLVMFuzzerInitialize(const uint8_t *buf, size_t len)
+{
+#ifdef S2N_TEST_IN_FIPS_MODE
+    S2N_TEST_ENTER_FIPS_MODE();
+#endif
+
+    GUARD(s2n_init());
+    GUARD(atexit(s2n_fuzz_atexit));
+    return 0;
+}
+
+/*
+ * Tokenize the input fuzz buffer based on NULL bytes into output array and return the number of tokens.
+ * Avoiding extra heap allocation here to increase fuzz test rate.
+ */
+size_t find_strings(const uint8_t *buf, size_t len, const char **output_strings, size_t max_strings)
```

Comment:
This is probably due to my unfamiliarity with how fuzz test input is generated, but how did you get many null characters in the input? Or is this a less common case?

Reply:
The assumption here is that because my test is reliant on NULL bytes being in the input to generate SAN strings and thus cover more code branches, the fuzz test will learn and generate inputs that match that requirement. This is reflected in the corpus that was generated for this test. For example, here's a corpus file for this test that embedded NULLs as expected: https://github.com/awslabs/s2n/blob/abc4c3e5f2e3352ca94768789d140a68e6022853/tests/fuzz/corpus/s2n_select_server_cert_test/9c139ef5428b482fcac9141de2fe90333bc7c13f